services/controller/src/plz/controller/db_storage.py

```diff
@@ -0,0 +1,60 @@
+import json
+import logging
+from abc import ABC, abstractmethod
+from typing import Set
+
+from redis import StrictRedis
+
+log = logging.getLogger(__name__)
+
+
+class DBStorage(ABC):
+    @abstractmethod
+    def store_start_metadata(self, execution_id: str, start_metadata: dict) \
+            -> None:
+        pass
+
+    def retrieve_start_metadata(self, execution_id: str) -> dict:
+        pass
+
+    def add_execution_id_for_user_and_project(
+            self, user: str, project: str, execution_id: str) -> None:
+        pass
+
+    def retrieve_execution_ids_for_user_and_project(
+            self, user: str, project: str) -> Set[str]:
+        pass
+
+
+class _RedisDBStorage(DBStorage):
+    def __init__(self, redis: StrictRedis):
+        super().__init__()
+        self.redis = redis
+
+    def store_start_metadata(
+            self, execution_id: str, start_metadata: dict) -> None:
+        self.redis.hset(
+            'start_metadata', execution_id, json.dumps(start_metadata))
+        # Store by user and project in sets for quick lookup
+        user = start_metadata['configuration']['user']
+        project = start_metadata['configuration']['project']
+        self.redis.sadd(f'execution_ids_for_user#{user}', execution_id)
+        self.redis.sadd(f'execution_ids_for_project#{project}', execution_id)
+
+    def retrieve_start_metadata(self, execution_id: str) -> dict:
+        start_metadata_bytes = self.redis.hget('start_metadata', execution_id)
+        if start_metadata_bytes is None:
+            raise ValueError(f'No start metadata available for {execution_id}')
+        start_metadata = str(start_metadata_bytes, 'utf-8')
+        return json.loads(str(start_metadata))
+
+    def retrieve_execution_ids_for_user_and_project(
+            self, user: str, project: str) -> Set[str]:
+        return {str(e, 'utf-8')
+                for e in self.redis.sinter([
+                    f'execution_ids_for_user#{user}',
+                    f'execution_ids_for_project#{project}'])}
+
+
+def get_redis_db_storage(redis: StrictRedis):
+    return _RedisDBStorage(redis)
```

Comment:
This function seems unnecessary. I think it's fine to have _configuration.py_ knowing the concrete type.

Reply:
Cool